### ISSUE TITLE
Support for chunked transfer encoding

### ISSUE THREAD

**Issue by oysteing** [cloudfoundry-community/terraform-provider-cloudfoundry#388]
The provider crashes when trying to upload bits for an application resource with binary loaded over HTTP chunked transfer encoding (Content-Length header is missing). This is important as Nexus Repository Manager 3 provides all assets with chunked transfer encoding.

### Steps to reproduce
1. You will need a sample.jar, which could be any file.
2. Start an nginx server with gzip compression --> chunked transfer encoding

`echo "gzip on;\ngzip_types application/java-archive;" > gzip.conf`
`docker run --rm -v $PWD/sample.jar:/usr/share/nginx/html/sample.jar:ro -v $PWD/gzip.conf:/etc/nginx/conf.d/gzip.conf:ro -p 80:80 nginx`

```
terraform {
  required_providers {
    cloudfoundry = {
      source = "cloudfoundry-community/cloudfoundry"
      version = "0.15.3"
    }
  }
}

data "cloudfoundry_space" "target" {
  name = "space"
  org_name = "org"
}

resource "cloudfoundry_app" "chunked-test" {
  name             = "chunked-test"
  space            = data.cloudfoundry_space.target.id
  path             = "http://localhost/sample.jar"
}
```

### Actual result
```
% terraform apply

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # cloudfoundry_app.chunked-test will be created
  + resource "cloudfoundry_app" "chunked-test" {
      + buildpack                  = (known after apply)
      + command                    = (known after apply)
      + disk_quota                 = (known after apply)
      + enable_ssh                 = (known after apply)
      + environment                = (sensitive value)
      + health_check_http_endpoint = (known after apply)
      + health_check_timeout       = (known after apply)
      + health_check_type          = "port"
      + id                         = (known after apply)
      + id_bg                      = (known after apply)
      + instances                  = 1
      + memory                     = (known after apply)
      + name                       = "chunked-test"
      + path                       = "http://localhost/sample.jar"
      + ports                      = (known after apply)
      + space                      = "e941d736-149d-47bd-aa24-c1f0293c96cf"
      + stack                      = (known after apply)
      + stopped                    = false
      + strategy                   = "none"
      + timeout                    = 60
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

cloudfoundry_app.chunked-test: Creating...
╷
│ Error: Plugin did not respond
│
│   with cloudfoundry_app.chunked-test,
│   on app.tf line 6, in resource "cloudfoundry_app" "chunked-test":
│    6: resource "cloudfoundry_app" "chunked-test" {
│
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain more details.
╵

Stack trace from the terraform-provider-cloudfoundry_v0.15.3 plugin:

panic: read tcp [::1]:51478->[::1]:80: use of closed network connection

goroutine 104 [running]:
github.com/terraform-providers/terraform-provider-cloudfoundry/cloudfoundry/managers/bits.BitsManager.UploadApp.func1()
	github.com/terraform-providers/terraform-provider-cloudfoundry/cloudfoundry/managers/bits/bitsmanager.go:226 +0x565
created by github.com/terraform-providers/terraform-provider-cloudfoundry/cloudfoundry/managers/bits.BitsManager.UploadApp
	github.com/terraform-providers/terraform-provider-cloudfoundry/cloudfoundry/managers/bits/bitsmanager.go:200 +0x450

Error: The terraform-provider-cloudfoundry_v0.15.3 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

Debug log: [terraform-provider-cloudfoundry_v0.15.3_debug.log](https://github.com/cloudfoundry-community/terraform-provider-cloudfoundry/files/8402825/terraform-provider-cloudfoundry_v0.15.3_debug.log)

### Expected result
Start nginx without gzip to disable chunked
`docker run --rm -v $PWD/sample.jar:/usr/share/nginx/html/sample.jar:ro -p 80:80 nginx`

```
% terraform apply

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # cloudfoundry_app.chunked-test will be created
  + resource "cloudfoundry_app" "chunked-test" {
      + buildpack                  = (known after apply)
      + command                    = (known after apply)
      + disk_quota                 = (known after apply)
      + enable_ssh                 = (known after apply)
      + environment                = (sensitive value)
      + health_check_http_endpoint = (known after apply)
      + health_check_timeout       = (known after apply)
      + health_check_type          = "port"
      + id                         = (known after apply)
      + id_bg                      = (known after apply)
      + instances                  = 1
      + memory                     = (known after apply)
      + name                       = "chunked-test"
      + path                       = "http://localhost/sample.jar"
      + ports                      = (known after apply)
      + space                      = "e941d736-149d-47bd-aa24-c1f0293c96cf"
      + stack                      = (known after apply)
      + stopped                    = false
      + strategy                   = "none"
      + timeout                    = 60
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

cloudfoundry_app.chunked-test: Creating...
cloudfoundry_app.chunked-test: Still creating... [10s elapsed]
cloudfoundry_app.chunked-test: Still creating... [20s elapsed]
cloudfoundry_app.chunked-test: Still creating... [30s elapsed]
cloudfoundry_app.chunked-test: Still creating... [40s elapsed]
cloudfoundry_app.chunked-test: Creation complete after 50s [id=e70e9af4-6cc3-4c68-b7d8-2734895bf9bc]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```

**Comment by oysteing** [comment]
We've created a fix for this issue, and it has been running in production for several months. Patch is attached. The approach is to download the source file locally to a temp file, which is deleted at the end. This requires, of course, sufficient free disk space. Is this a possible way forward?

```
Index: cloudfoundry/managers/bits/bitsmanager.go
===================================================================
diff --git a/cloudfoundry/managers/bits/bitsmanager.go b/cloudfoundry/managers/bits/bitsmanager.go
--- a/cloudfoundry/managers/bits/bitsmanager.go	(revision 07622837a1f843d72332536e8e0741b1ce251d1f)
+++ b/cloudfoundry/managers/bits/bitsmanager.go	(date 1655459421517)
@@ -321,11 +321,32 @@
 				baseName = params["filename"]
 			}
 		}
-		return ZipFile{
-			r:        resp.Body,
-			baseName: baseName,
-			filesize: fileSize,
-		}, nil
+		if fileSize > 0 {
+			return ZipFile{
+				r:        resp.Body,
+				baseName: baseName,
+				filesize: fileSize,
+			}, nil
+		} else {
+			tempFile, err := ioutil.TempFile("", "")
+			if err != nil {
+				return ZipFile{}, err
+			}
+			defer os.Remove(tempFile.Name())
+			fileSize, err := io.Copy(tempFile, resp.Body)
+			if err != nil {
+				return ZipFile{}, err
+			}
+			_, err = tempFile.Seek(0, 0)
+			if err != nil {
+				return ZipFile{}, err
+			}
+			return ZipFile{
+				r:        tempFile,
+				baseName: baseName,
+				filesize: fileSize,
+			}, nil
+		}
 	}
 	f, err := os.Open(path)
 	if err != nil {
```

**Comment by loafoe** [comment]
@oysteing nice! Can you create a Pull Request with this fix against the `master` branch?

**Comment by oysteing** [comment]
@loafoe Sure: https://github.com/cloudfoundry-community/terraform-provider-cloudfoundry/pull/402

**Comment by loafoe** [comment]
@oysteing available in `v0.15.5`, thanks for fixing!
